### PR TITLE
Added handling for csrftoken

### PR DESCRIPTION
**Pull request by mkmitchell**
First pull request!

For some reason there was an extra space in some of the code so I removed those in the area I was working in.  I'm not sure if this works for proxy but it worked fine for no_proxy.

### CODE REVIEW

**Comment by olivierhagolle** [comment]
Hi, sorry, but I have not been able to make it work.
No file is downloaded, I just get a html dump which is a new login screen. But my account works when provided online
Olivier

**Comment by mkmitchell** [comment]
Hey, I've been getting that html dump as well before this error started
happening.  I had to add a sys.exit() after a product wasn't found or it
would just keep looping the dump.

Does it look something like this?
"
python F:\Landsat\download_landsat_scene.py -o catalog -z -b LC8 -u
usgs.txt -d 20160201 -f 20160331 --output F:\Landsat\Download
--outputcatalogs=F:\Landsat\Download -c 50 -s 028035
None None
Verifying catalog metadata files...
Searching for images in catalog...
erreur : le fichier est au format html
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "
http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><html xmlns="
http://www.w3.org/1999/xhtml" xml:lang="en" lang="en" dir="ltr">
    <head>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8"
/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="description" content="Authentication Portal" />
<meta name="keywords"
    content="" />
<meta name="author" content="USGS - U.S. Geological Survey" />
<meta name="publisher" content="USGS - U.S. Geological Survey" />
<meta name="classification" content="Government" />
<meta name="robots" content="ALL" />
<meta name="distribution" content="Global" />
<meta name="rating" content="Safe For Kids" />
<meta name="copyright" content="None" />
<meta name="language" content="English" />
<meta name="doc-type" content="Government" />
<meta name="doc-class" content="Living Document" />
<meta name="doc-rights" content="Public Domain" />

<link href="https://ers.cr.usgs.gov/css/jquery-ui-1.11.2.min.css"
rel="stylesheet" type="text/css" />
<link href="
https://maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css"
rel="stylesheet" />
<link href="https://ers.cr.usgs.gov/css/base.css" rel="stylesheet"
type="text/css" />
<link href="https://ers.cr.usgs.gov/css/dropdownmenu.css" rel="stylesheet"
type="text/css" />

<script type="text/javascript" src="
https://ajax.googleapis.com/ajax/libs/jquery/2.1.1/jquery.min.js"></script>
<script type="text/javascript" src="
https://ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/jquery-ui.min.js"
></script>
<script type="text/javascript" src="
https://ers.cr.usgs.gov/js_library/sticky.min.js"></script>
<script type="text/javascript" src="
https://ers.cr.usgs.gov/js_include/ERS.js"></script>        <title>EROS
Registration System (ERS)</title>        <link href="/css/login.css"
media="screen" rel="stylesheet" type="text/css" />            </head>
    <body>
        
<div id="header">
    <div id="headerLogo">
        <a href="http://www.usgs.gov/" target="_blank">
            USGS - science for a changing world
        </a>
    </div>

    <div id="headerGradientLeft"></div>

    <div id="usgsInfo">
        <a href="http://www.usgs.gov/" target="_blank">USGS Home</a><br />
        <a href="http://www.usgs.gov/ask/" target="_blank">Contact
USGS</a><br />
        <a href="http://search.usgs.gov/" target="_blank">Search USGS</a>
    </div>

    <div id="headerGradientRight"></div>

    <div style="clear:both;"></div>
</div>

        <div id="titleBar">
            EROS Registration System (ERS)        </div>


        <div id="pageContent">
            <script type="text/javascript">
    $(document).ready(function()
    {
        $('#loginButton').button();

        $('.oAuthType').click(function()
        {
                window.location = 'https://ers.cr.usgs.gov/login/oauth/' +
$(this).attr('data-serviceKey');
        });

        $('#registerButton').button().click(function()
        {
                document.location = 'https://ers.cr.usgs.gov/register';
        });

        $('#username').focus();
    });
</script>

<div id="loginHeading">
            <p>ERS consolidates user profile and authentication for all
EROS web services into a single independent application.</p>
        </div>
<div id="centeredContainer">

        <h2>Sign In</h2>
                <div class="seperator">
                <span></span>
                sign in with your existing USGS registered username and
password
                <span></span>
        </div>
        <form id="loginForm" name="loginForm"
enctype="application/x-www-form-urlencoded" action="/login/" method="post">
              <div>
                        <div class="inputContainer">

<input type="text" name="username" id="username" value=""
placeholder="Registered USGS Username" class="textInput" maxlength="30" />
                                                   </div>
                        <div class="inputContainer">

<input type="password" name="password" id="password" value=""
placeholder="Registered USGS Password" class="textInput" />
                                      </div>
                                        </div>
                <div class="signinOptions clearfix">
                        <div id="forgotPasswordContainer">
                                <a href="
https://ers.cr.usgs.gov/password/request">forgot password?</a>
                        </div>
                </div>
                <dt id="csrf_token-label">&#160;</dt>
<dd id="csrf_token-element">
<input type="hidden" name="csrf_token" value="lMx7JQ==" id="csrf_token"
/></dd>         <input id="loginButton" type="submit" value="Sign In" />
        </form>
        <div class="seperator">
                <span></span>
                Don't have an account?
                <span></span>
        </div>
        <input type="button" id="registerButton" value="Create New Account"
/>
</div>
</br>

<div id="praMessage">
    OMB number 1028-0119</br>
    OMB expiration date 06/30/2019</br><br>
ΓÇï    Privacy and Paperwork Reduction Act statements: 16 U.S.C. 1a7
authorized collection of this
    information. This information will be used by the U.S. Geological
Survey to better serve the public.
    Response to this request is voluntary.  No action may be taken against
you for refusing to supply the
    information requested.  The time required to complete this information
collection is estimated to
    average 5 minutes per response.  We will not distribute responses
associated with you as an individual.
    We ask you for some basic organizational and contact information to
help us interpret the results and,
    if needed, to contact you for clarification.  Comments on this
collection should be sent to
    <a href="mailto:custserv@usgs.gov">custserv@usgs.gov</a>.ΓÇï
</div>



        </div>
        
<div id="footer">
    <div id="usgsfooter">
        <p id="usgsfooterbar" class="clearfix">
            <a class="alignLeft" target='new' title="Accessibility Policy
(Section 508)" href="http://www.usgs.gov/laws/accessibility.html
">Accessibility</a>
            <a class="alignLeft" target='new' title="Freedom of Information
Act" href="http://www.usgs.gov/foia/">FOIA</a>
            <a class="alignLeft" target='new' title="Privacy policies of
the U.S. Geological Survey." href="http://www.usgs.gov/laws/privacy.html
">Privacy</a>
            <a class="alignLeft" target='new' title="Policies and notices
that govern information posted on USGS Web sites." href="
http://www.usgs.gov/laws/policies_notices.html">Policies and Notices</a>

            <a class="alignRight" target='help' href="
https://lta.cr.usgs.gov/ee_ers">Help</a>
            <a class="alignRight" target='feedback' href="
https://ers.cr.usgs.gov/feedback">Feedback</a>
        </p>
        <p id="usgsfootertext">
            <a href="http://usa.gov/">
                USA.gov logo
            </a>
            <a href="http://www.doi.gov/">U.S. Department of the
Interior</a>
            <a href="http://www.usgs.gov/">U.S. Geological Survey</a>
            <br/>
            URL: https://ers.cr.usgs.gov            <br/>
            Page Contact Information:
            <a href="mailto:custserv@usgs.gov">custserv@usgs.gov</a>
            <br/>
            Page Last Modified:
            08/09/2016                    </p>
    </div>
</div>


<script async type="text/javascript" src="
https://www2.usgs.gov/scripts/analytics/usa-search.js"></script>
<script type="text/javascript" src="
https://www.google-analytics.com/analytics.js"></script>

<script type="text/javascript">
    try {
        ga('set', 'anonymizeIp', true);

        ga('create', 'UA-21295076-2', 'auto', {'name': 'lta'});

        //Assuming this is what will be required once usgs analytics is
updated
        ga('lta.send', 'pageview');
    } catch (exception) {
        console.log(exception);
    }
</script>

<script type="text/javascript">
    $(document).ajaxSend(function(e,request,settings){
        try {
            //Assuming this is what will be required once usgs analytics is
updated
            ga('lta.send', 'pageview', {'location': settings.url});
        } catch (exception) {
            console.log(exception);
        }
    });
</script>    </body>
</html>

   product LC80280352016035LGN00 not found
"

**Comment by mkmitchell** [comment]
I tested one that I knew should work and got the same dump.  I'll work on this in a bit.  A few meetings first.

**Comment by olivierhagolle** [comment]
yes, the tool provides the dump if it gets an html file instead of a LANDSAT file.
Means it is not working.

I have replaced the dump by a log of the html file in an error file which is more convenient. 
You might want to replace the block from  if (req.info().gettype()=='text/html'): by

`if (req.info().gettype()=='text/html'):
      print "error : file is in html format, and shouldn't"
      lignes=req.read()
      if lignes.find('Download Not Found')>0 :
            raise TypeError
      else:
          with open("error_output.html","w") as f:
              f.write(lignes)
              print "result saved in ./error_output.html"
              print sys.exit(-1)`

**Comment by mkmitchell** [comment]
I'm pretty new at all this fun web stuff.
It appears I'm not getting the token correctly.  I'll keep working on it.

**Comment by mkmitchell** [comment]
I'm not sure if this will help you but I successfully requested the website, got a cookie PHPSESSID then sent another request but still get the 403 error.  I pulled the other cookie information from fiddler and copied it in with no luck.

`
 log_data = {
        'username': usgs['account'],
        'password': usgs['passwd']
    }
    post_data = urllib.urlencode(log_data)

    cookjar = cookielib.CookieJar()
    opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(cookjar))
    headers = {'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
            'Accept-Language': 'en-US,en;q=0.5',
            'User-Agent': 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:44.0) Gecko/20100101 Firefox/44.0',
            'Host': 'ers.cr.usgs.gov:443',
            'Connection': 'keep-alive'}
    req = urllib2.Request("https://ers.cr.usgs.gov/login")
    f = opener.open(req)
    data = dict((cookie.name, cookie.value) for cookie in cookjar)
    headers = {'Host': 'ers.cr.usgs.gov',
        'Connection': 'keep-alive',
        'Cache-Control': 'max-age=0',
        'Upgrade-Insecure-Requests': '1',
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/53.0.2780.0 Safari/537.36',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
        'Accept-Encoding': 'gzip, deflate, sdch, br',
        'Accept-Language': 'en-US,en;q=0.8',
        'Cookie:': 'PHPSESSID='+data["PHPSESSID"]}
    headers = {'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        'Accept-Language': 'en-US,en;q=0.5',
        'User-Agent': 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:44.0) Gecko/20100101 Firefox/44.0',
        'Host': 'ers.cr.usgs.gov',
        'Connection': 'keep-alive',
        'Accept-Encoding': 'gzip, deflate, sdch, br',
        'Cookie:': 'PHPSESSID='+data["PHPSESSID"] + '; ee-system-notices=%5B%225261%22%2C%225141%22%5D; _ga=GA1.2.1349842508.1469734657; _gat_lta=1'}
    req = urllib2.Request("https://ers.cr.usgs.gov/login", post_data, headers=headers)
    f = opener.open(req)
`

**Comment by mkmitchell** [comment]
Alright.  I fixed it for real this time.

**Comment by mkmitchell** [comment]
Alright.  I had to add the need for beautifulsoup because the way I got the csrf_token was to parse the html.  Please hack away at it to make it meet your needs.  I was exploring grabbing headers from the site and passing them in.  The odd thing is that the headers I'm passing is completely empty but if I remote sending them it won't work.

**Comment by mkmitchell** [comment]
I committed again and removed the header stuff and just send headers={}.  This removed the need for itertools.

**Comment by olivierhagolle** [comment]
Great, it works !
The only drawback is that we have to download BeautifulSoup.
Thanks a lot Mike.
Olivier
